File: client/web/app/technical-indicators/ema.service.ts
import { StockPrice } from '../charts/stock-price'
import { EMA } from './ema'

// http://investexcel.net/how-to-calculate-ema-in-excel/
export class EMAService {
    PARAM_1: number = 10;
    calculate(stockPrices: StockPrice[]) : EMA[] {
        //debugger;
        let entries = stockPrices.map(sp => new EMA(sp.date, sp.price));
        entries.reduce((previousValue, currentValue, currentIndex) => {
            let index = currentIndex + 1;
            currentValue.priceSum = previousValue.priceSum + currentValue.price;
            currentValue.priceAverage = currentValue.priceSum / index;
            
            // 10 Day EMA
            if (index === this.PARAM_1) {
                currentValue.ema = currentValue.priceAverage;
            } else if (index > this.PARAM_1) {
                currentValue.ema = currentValue.price * (2/(this.PARAM_1 + 1)) + previousValue.ema * (1 - 2/(this.PARAM_1 + 1));
            }

            return currentValue;
        }, new EMA(new Date(1900, 1, 1), 0));

        return entries;
    }
}

File: server/RuleOne/src/InvSys.RuleOne.Core/Services/ThreeTools/EMAService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using InvSys.RuleOne.Core.Models.ThreeTools;
using InvSys.StockQuotes.Api.Client.Proxy.Models;

namespace InvSys.RuleOne.Core.Services.ThreeTools
{
    // EMA - Exponential Moving Average
    // http://investexcel.net/how-to-calculate-ema-in-excel/
    public class EMAService : IEMAService
    {
        public ICollection<EMAData> Calculate(ICollection<HistoricalQuote> prices, int days)
        {
            var entries = prices.Select(p => new EMAData { Date = p.Date.Value, Price = (decimal)p.Close.Value }).ToList(); // TODO: swagger and decimal (https://github.com/Azure/autorest/issues/83)
            entries.Insert(0, new EMAData { Date = new DateTime(1900, 1, 1), Price = 0 });
            for (var index = 1; index < entries.Count; index++)
            {
                var previousValue = entries[index - 1];
                var currentValue = entries[index];

                currentValue.PriceSum = previousValue.PriceSum + currentValue.Price;
                currentValue.PriceAverage = currentValue.PriceSum / index;

                // x Days EMA
                if (index == days)
                {
                    currentValue.EMA = currentValue.PriceAverage;
                }
                else if (index > days)
                {
                    currentValue.EMA = currentValue.Price * ((decimal)2 / (days + 1)) + previousValue.EMA * (1 - (decimal)2 / (days + 1));
                }
            }

            return entries.Where(e => e.EMA != 0).OrderBy(e => e.Date).ToList();
        }
    }
}
